include/exadg/fluid_structure_interaction/precice/coupling_surface.h

```diff
@@ -0,0 +1,235 @@
+#pragma once
+
+#include <deal.II/base/conditional_ostream.h>
+#include <deal.II/base/vectorization.h>
+
+#include <deal.II/matrix_free/matrix_free.h>
+
+#include <exadg/matrix_free/integrators.h>
+#include <precice/SolverInterface.hpp>
+
+namespace ExaDG
+{
+namespace preCICE
+{
+using namespace dealii;
+
+/**
+ * Enum to handle all implemented data write methods one can use
+ */
+enum class WriteDataType
+{
+  undefined,
+  values_on_dofs,
+  values_on_other_mesh,
+  gradients_on_other_mesh,
+  values_on_q_points,
+  normal_gradients_on_q_points
+};
+
+/**
+ * A pure abstract base class, which defines the methods for the functions
+ * used in the main Adapter class. Each instance of all derived classes are
+ * always dedicated to a specific coupling mesh, i.e., the vertices used for
+ * the coupling. The instantiated objects provide functions on how to read
+ * and write data on this mesh and how to define the mesh by means of its
+ * spatial coordinates.
+ */
+template<int dim, int data_dim, typename VectorizedArrayType>
+class CouplingSurface
+{
+public:
+  CouplingSurface(std::shared_ptr<const MatrixFree<dim, double, VectorizedArrayType>> data,
```

Comment:
This might be possible, but it is not implemented and requires additional effort in other places such as the global data vectors and the actual data passing to preCICE. What would be the use case? Keep in mind that this change would only make sense in case the other participant ('the other side') also uses single precision.

I don't have a use case for this at the moment. If you think this change would be valuable, I would put it on the future TODO list.

Reply:
Sounds fine!